nixos/modules/misc/wordlist: init

OmnipotentEntity opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Addresses #16545.  Allows for user defined environment variables that hold paths to wordlists.  This is to allow for easy access to wordlists for users and scripts, (in other distributions a convenient wordlist is typically found in /usr/share/dict/words or similar).  The default wordlist is the one found in `scowl`, for no other reason than that's the one that was mentioned in the linked issue.  Further work is required for patching whichever programs use these wordlists.

It is possible to specify multiple environment variables as well.  This is for users who need multiple wordlists (such as multilingual users).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

OmnipotentEntity committed:
nixos/modules/misc/wordlist: init

Addresses #16545.  Allows for user defined environment variables that
hold paths to wordlists.  This is to allow for easy access to wordlists
for users and scripts, (in other distributions a convenient wordlist is
typically found in /usr/share/dict/words or similar).  The default
wordlist is the one found in scowl, for no other reason than that's the
one that was mentioned in the linked issue.

It is possible to specify multiple environment variables as well.  This
is for users who need multiple wordlists (such as multilingual users).

OmnipotentEntity commented:
Notes:

I have not tested this as part of nixpkgs, I have only tested this file as an included module in my system flake.  It seems to work, but also seems to require a fresh login.

I have no idea where this module should go, so I threw it in misc/.  I didn't think it belonged in either programs or services, because it is neither a program nor a service.

This is my first nixos module, please let me know if I did anything outrageously wrong and I'll correct it.

Synthetica9 commented:
Maybe we could even merge multiple lists? If they are just simple lists of words, I don't see why we couldn't

OmnipotentEntity commented:
> Maybe we could even merge multiple lists? If they are just simple lists of words, I don't see why we couldn't

I would imagine it would be better to handle that particular task outside of this module (through another derivation) and then just reference that derivation using the module.

Synthetica9 commented:
Doing it in the module has the advantage that multiple wordlists can be set
and it will Just Work™️, even if they are set in multiple places.

On Fri, 4 Feb 2022, 17:44 OmnipotentEntity, ***@***.***>
wrote:

> Maybe we could even merge multiple lists? If they are just simple lists of
> words, I don't see why we couldn't
>
> I would imagine it would be better to handle that particular task outside
> of this module (through another derivation) and then just reference that
> derivation using the module.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/158034#issuecomment-1030159040>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABV7PJ4ZGT4HNKH3JU4EN5LUZP65ZANCNFSM5NP5JJNQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

OmnipotentEntity commented:
If that's the case then the file resulting from the concatenation will have to be stored somewhere in the nix store, but I was under the impression that a module typically does not have its own presence in the nix store.  It'd be part of the larger derivation.  Do you have an idea what that might look like or how that might work?

OmnipotentEntity commented:
The failing bit is documentation failing because I have included `${pkgs.scowl}` in the default.  I can simply remove it, of course, but is there was a way to make the documentation report a literal `${pkgs.scowl}` for the default, rather than its expanded version?

rissson commented:
> I was under the impression that a module typically does not have its own presence in the nix store. It'd be part of the larger derivation.

That's actually what happens when you set something via `environment.etc`, so I guess that's fine.

rissson commented:
I'm not sure that creating a top-level module `wordlist` is the way to go. Maybe it's best scoped under one of the existing ones, perhaps `environment`.

OmnipotentEntity commented:
> I'm not sure that creating a top-level module `wordlist` is the way to go. Maybe it's best scoped under one of the existing ones, perhaps `environment`.

Sounds good to me.

The force push is not quite ready yet.  There's still a major problem with disambiguating between strings and paths.  I am trying to work it out.  So I've converted this to a draft.

pennae commented:
why not have it accept *only* paths, and have users who wish to use nix strings as a wordlist turn them into a path with writeFile themselves? presumably wordlists warrant being put into a dedicated file anyway, so that shouldn't be much of a loss.

pennae commented on nixos/modules/misc/wordlist.nix:
```diff
@@ -0,0 +1,116 @@
+    };
+
+    words = mkOption {
+      type = types.nullOr types.nonEmptyListOf ( types.str );
```
```suggestion
      type = types.nullOr (types.nonEmptyListOf types.str);
```

pennae reviewed (commented):
almost all options are missing descriptions, those should definitely be added.

sorry for the barrage of comments on such a small module; most of them are minor nitpicks but some are substantial. we should endeavor to make module behaviour as unsurprising as possible, and currently this module can exhibit some quite surprising behavior.

OmnipotentEntity commented on nixos/modules/misc/wordlist.nix:
```diff
@@ -0,0 +1,116 @@
+processOne = source:
+  if source ? file
+  then
+    builtins.readFile source.file
```
IFD?

pennae replied:
import from derivation. if you supply a path from a package as `source.file` that package has to be built during the evaluation phase, which is a single-threaded process and thus slows down a lot. for that reason (and others) it's disallowed in some installations, so it's best to avoid doing things that require nix to read files from derivation outputs during eval.

OmnipotentEntity commented on nixos/modules/misc/wordlist.nix:
```diff
@@ -0,0 +1,116 @@
+      enable = mkEnableOption "wordlist environment variables.";
+
+      lists = mkOption {
+        type = types.nonEmptyListOf ( specType );
```
I'm not quite sure I understand!  But I've made the requested change.

pennae commented on nixos/modules/misc/wordlist.nix:
```diff
@@ -0,0 +1,50 @@
+        WORDLIST = [ "${pkgs.scowl}/share/dict/words.txt" ];
+      };
+
+      defaultText = literalExpression ''{
+        WORDLIST = [ "''${pkgs.scowl}/share/dict/words.txt" ];
+      };
+      '';
```
```suggestion
      defaultText = literalExpression ''
        {
          WORDLIST = [ "''${pkgs.scowl}/share/dict/words.txt" ];
        }
      '';
```
leading common whitespace is removed in `''`-quoted strings, but this one has none. the string as written looks like this when printed:
```
{
        WORDLIST = [ "''${pkgs.scowl}/share/dict/words.txt" ];
      };
```
which isn't usually the intention. the same goes for other `''` strings in this module.

here the trailing `;` is also misplaced here since it's not a part of the expression, rather it's part of an attribute binding.

OmnipotentEntity replied:
oh dang, thanks for the catch, I thought I had fixed all of these errors after the last time you pointed this out.

pennae commented on nixos/modules/misc/wordlist.nix:
```diff
@@ -0,0 +1,50 @@
+processList = name: paths:
+  if builtins.length paths == 1
+  then
+    head paths
```
this will probably not work if a user sets eg `environment.worlists.WORDLIST = [ ./wordlist ]` because paths themselves aren't strings. both legs if this condition will need a `toString`, so it might make sense to move the toString into the mapAttrs below instead.

OmnipotentEntity replied:
> did the enable option go away? it would be very good to have one for this module, scowl would add over 100MB to our system closure size for no tangible benefit to us.
> 
> indentation of stuff in the top-level let is also a bit wonky, usually let bindings are indented by two spaces. but all in all this is shaping up nicely :)

Sorry, I removed the `enable` option, because it seemed redundant to have "enable this" along with "use this list of environment variables".  I should have also emptied the default as well, but it slipped my mind.

pennae reviewed (commented):
did the enable option go away? it would be very good to have one for  this module, scowl would add over 100MB to our system closure size for no tangible benefit to us.

indentation of stuff in the top-level let is also a bit wonky, usually let bindings are indented by two spaces. but all in all this is shaping up nicely :)

OmnipotentEntity commented:
> did the enable option go away? it would be very good to have one for this module, scowl would add over 100MB to our system closure size for no tangible benefit to us.
> 
> indentation of stuff in the top-level let is also a bit wonky, usually let bindings are indented by two spaces. but all in all this is shaping up nicely :)

Sorry, I removed the `enable` option, because it seemed redundant to have "enable this" along with "use this list of environment variables".  I should have also emptied the default as well, but it slipped my mind.  That being said, do you think there's a good use case for having an enable option?  I couldn't think of one, but admittedly I'm not very familiar with complex nix workflows.

pennae commented:
for a module like this it could go either way. if the default is `{}` but the example lists how to get a good default wordlist then an enable option might not be necessary, but then again if the default is applicable to almost everyone an enable option is easier to use. if that's the case it may even make sense to pare the module down even further, so that it only sets `WORDLIST` (ie, provides options `environment.wordlist.enable` and `environment.wordlist.sources`).

really depends on how extensible this needs to be to fit most people. personally i'd guess that the last option here (enable + sources) is sufficient, but we don't use wordlists at all so that's not an opinion worth a whole lot.

OmnipotentEntity commented:
Put back the default, adjusted the indentation on the `let` block, fixed the `literalExpression` spacing, and moved `toString` to the correct location.

pennae commented on nixos/modules/misc/wordlist.nix:
```diff
@@ -0,0 +1,65 @@
+  config = mkIf config.environment.wordlist.enable {
+    environment.variables =
+      lib.mapAttrs
+        (name: value: builtins.toString (processList "wordlist-${name}" value))
```
we shouldn't write reviews so close to bedtime :sweat: `toString` is of course not sufficient because paths don't have context, so `toString ./list == "/path/to/the/list"` and nothing is copied into the store. we actually need
```suggestion
        (name: value: "${processList "wordlist-${name}" value}")
```
to copy contents of path literals into the store so that all users can access them. in fact we may even want to use `concatText` in all cases, even lists of length 1, because that doesn't require us keeping the full scowl package around: the word list gets copied out during system build, and scowl can be GCd afterwards to get rid of parts of it that aren't actually in use.

pennae replied:
it's okay to have this behaviour if the user put a stringy path into the config (eg `[ "/etc/words" ]` or so). using path literals should always copy the referenced file into the store though, since that's what's generally expected to happen when using path literals. :)

OmnipotentEntity commented:
Alright, I've made those changes. I've also cleaned up the documentation a little bit.  Really happy with how this module is turning out, finally.

pennae commented on nixos/modules/misc/wordlist.nix:
```diff
@@ -0,0 +1,52 @@
+            AUGMENTED_WORDLIST = [
+              "''${pkgs.scowl}/share/dict/words.txt"
+              "''${pkgs.scowl}/share/dict/words.variants.txt"
+              "''${pkgs.writeText "extra-words" (lib.concatStringsSep "\n" [ "desynchronization" "oobleck" ""])}"
+              # Final empty string to get a trailing newline, important for composing multiple lists.
```
```suggestion
              (builtins.toFile "extra-words" ''
                desynchronization
                oobleck
              '')
```

OmnipotentEntity replied:
> we just tested simple concatenation of wordlists for `look` (which seems the main use case in the linked issue?), and it doesn't work. confused `look` requires the word list to be sorted to function, so this module has to sort its word lists as well. (and if we have to use more complicated methods of building the final wordlist we can also insert a `\n` between two files, obviating the need for a trailing `\n` on every file)
> 
> apart from that this module does look good now though.

Sorting means that I can't use the built-in `pkgs.concatText` but that's OK.  I'll just go back to `pkgs.runCommand` instead, and add a few better ways to make sure that the whitespace is correct (and there aren't doubled entries and so on).

OmnipotentEntity replied:
> we just tested simple concatenation of wordlists for `look` (which seems the main use case in the linked issue?), and it doesn't work. confused `look` requires the word list to be sorted to function, so this module has to sort its word lists as well. (and if we have to use more complicated methods of building the final wordlist we can also insert a `\n` between two files, obviating the need for a trailing `\n` on every file)
> 
> apart from that this module does look good now though.

No worries, I just think this means I can't use `pkgs.concatText` anymore, but that's not a big deal.  It also gives me the chance to make damn sure that all of the whitespace is right, and there's no leading/trailing space and there's no empty lines.  `awk 1` is there to concatenate the files while adding newlines between files if needed.  Fewer footguns.

pennae reviewed (commented):
we just tested simple concatenation of wordlists for `look` (which seems the main use case in the linked issue?), and it doesn't work. :confused:  `look` requires the word list to be sorted to function, so this module has to sort its word lists as well. (and if we have to use more complicated methods of building the final wordlist we can also insert a `\n` between two files, obviating the need for a trailing `\n` on every file)

apart from that this module does look good now though.

OmnipotentEntity commented:
> we just tested simple concatenation of wordlists for `look` (which seems the main use case in the linked issue?), and it doesn't work. confused `look` requires the word list to be sorted to function, so this module has to sort its word lists as well. (and if we have to use more complicated methods of building the final wordlist we can also insert a `\n` between two files, obviating the need for a trailing `\n` on every file)
> 
> apart from that this module does look good now though.

No worries, I just think this means I can't use `pkgs.concatText` anymore, but that's not a big deal.  It also gives me the chance to make damn sure that all of the whitespace is right, and there's no leading/trailing space and there's no empty lines.  `awk 1` is there to concatenate the files while adding newlines between files if needed.  Fewer footguns.

Synthetica9 commented:
Maybe also `sort | uniq` it to remove duplicates?

OmnipotentEntity commented:
Already doing that :)

OmnipotentEntity commented:
Ok, I think I'm done messing with this.  The final changes were documentation changes to better explain that this process is somewhat more robust than a simple concatenation.

Synthetica9 commented:
Thank you for your patience! Looks great to me now.
